### ISSUE TITLE
Allow to reference data from index (of pandas dataframes)

### ISSUE THREAD

**Issue by fhg-isi** [altair-viz/altair#3331]
Currently altair only seems to alllow referencing named columns by design:
https://altair-viz.github.io/user_guide/data.html#including-index-data

Please consider to support indexed pandas dataframes in a future altair version. Also see

https://stackoverflow.com/questions/77993730/how-to-use-indexed-data-frames-with-altair/

**Comment by binste** [comment]
Thanks for the suggestion. I see that it's easier to not have to write `.reset_index()`. The challenge is that Altair would need to call `.reset_index()` internally for every Pandas dataframe to make the index accessible. In many cases where the index is not needed for the chart, it would lead to unnecessary data being added to the Vega-Lite specification, see [Altair Internals](https://altair-viz.github.io/user_guide/internals.html).

For this to work, Altair would need to know when the index is used and when it isn't so it can call `.reset_index` in only those cases. Altair can only do this with the help of [VegaFusion](https://vegafusion.io/), see https://github.com/altair-viz/altair/issues/2428 for details on why.

In short, I think it adds a lot of complexity to make this work and it would only work with VegaFusion which is an additional dependency. I'll leave this open in case I'm missing something.

**Comment by fhg-isi** [comment]
* Index could be accessed without resetting the data_frame  (e.g. df.index.values) ?
*  Type of index could be checked ? If its not RangeIndex, consider it as explicit index ?
* Reset could only be called in the required cases where some expression like "$index" is used ?
* Spend extra method or option for indexed dataframes (use_index=True) ) ?
* Allow index as type for mapping `x = df.index` ?

Also see

https://stackoverflow.com/questions/20084487/use-index-in-pandas-to-plot-data

```
import pandas as pd

df = pd.DataFrame(
    [
        {'id_foo': 1, 'energy_carrier': 'oil', '2000': 5, '2020': 10},
        {'id_foo': 2, 'energy_carrier': 'electricity', '2000': 10, '2020': 20},
    ]
)

print(type(df.index))   # <class 'pandas.core.indexes.range.RangeIndex'>

indexed_df = df.pivot_table(
    columns='energy_carrier',
    values=['2000', '2020'],
    aggfunc='sum',
)

print(type(indexed_df.index))  # <class 'pandas.core.indexes.base.Index'>

df.set_index('id_foo', inplace=True)  # <class 'pandas.core.indexes.numeric.Int64Index'>

print(type(df.index))
```

**Comment by joelostblom** [comment]
Closing as the solution to doing this automatically seems more complex than writing out reset_index() each time.
